libgeos/libgeos_handle.go

```diff
@@ -0,0 +1,236 @@
+	"fmt"
+	"strings"
+	"unsafe"
+
+	"github.com/peterstace/simplefeatures/geom"
+)
+
+// Handle is an opaque handle that can be used to invoke libgeos operations.
+// Instances are not threadsafe and thus should not be used in a concurrent
+// manner.
+type Handle struct {
+	context C.GEOSContextHandle_t
+	errBuf  [1024]byte
+}
+
+// NewHandle creates a new libgeos handle.
+func NewHandle() (*Handle, error) {
+	h := &Handle{}
+	h.context = C.sf_init(unsafe.Pointer(&h.errBuf))
+	if h.context == nil {
+		return nil, errors.New("could not create libgeos context")
+	}
+	return h, nil
+}
+
+// err gets the last error message reported by libgeos as an error type. It
+// always returns a non-nil error. If no error message has been reported, then
+// it returns a generic error message.
+func (h *Handle) err() error {
+	msg := h.errMsg()
+	if msg == "" {
+		// No error stored, which indicates that the error handler didn't get
+		// trigged. The best we can do is give a generic error.
+		msg = "libgeos internal error"
+	}
+	h.errBuf = [1024]byte{} // Reset the buffer for the next error message.
+	return errors.New(strings.TrimSpace(msg))
+}
+
+// errMsg gets the textual representation of the last error message reported by
+// libgeos.
+func (h *Handle) errMsg() string {
+	// The error message is either NULL terminated, or fills the entire buffer.
+	firstZero := len(h.errBuf)
+	for i, b := range h.errBuf {
+		if b == 0 {
+			firstZero = i
+			break
+		}
+	}
+	return string(h.errBuf[:firstZero])
+}
+
+// Release releases any resources held by the handle. The handle should not be
+// used after Release is called.
+func (h *Handle) Release() {
+	C.GEOS_finish_r(h.context)
+}
+
+// createGeometryHandle converts a Geometry object into a libgeos geometry handle.
+func (h *Handle) createGeometryHandle(g geom.Geometry) (*C.GEOSGeometry, error) {
+	wkbReader := C.GEOSWKBReader_create_r(h.context)
+	if wkbReader == nil {
+		return nil, h.err()
+	}
+	defer C.GEOSWKBReader_destroy_r(h.context, wkbReader)
+
+	wkb := new(bytes.Buffer)
+	if err := g.AsBinary(wkb); err != nil {
+		return nil, err
+	}
+	gh := C.GEOSWKBReader_read_r(
+		h.context,
+		wkbReader,
+		(*C.uchar)(&wkb.Bytes()[0]),
+		C.ulong(wkb.Len()),
+	)
+	if gh == nil {
+		return nil, h.err()
+	}
+	return gh, nil
+}
+
+// ErrGeometryCollectionNotSupported indicates that a GeometryCollection was
+// passed to a function that does not support GeometryCollections.
+var ErrGeometryCollectionNotSupported = errors.New("GeometryCollection not supported")
+
+// Equals returns true if and only if the input geometries are spatially equal.
+func (h *Handle) Equals(g1, g2 geom.Geometry) (bool, error) {
+	if g1.IsEmpty() && g2.IsEmpty() {
+		// Part of the mask is 'dim(I(a) ∩ I(b)) = T'.  If both inputs are
+		// empty, then their interiors will be empty, and thus
+		// 'dim(I(a) ∩ I(b) = F'. However, we want to return 'true' for this
+		// case. So we just return true manually rather than using DE-9IM.
+		return true, nil
```

Comment:
Is it doing this because libgeos does the same way?

Reply:
Yes, that's correct. We have to do this because `Equals` and `Relate("T*F**FFF*")` aren't _exactly_ the same. Libgeos _does_ do this sort of logic for its `Equals` method as well (and Postgis does something very similar for `ST_Equals` as well).

I _could have_ wrapped libgeos `Equals` method instead of just wrapping the `Relate` method. However, my idea is that if we only wrap the `Relate` method, then we can later switch it out once we build the DE-9IM engine in Go.